File: app/src/main/java/pt/cosmik/boostctrl/external/tournament_brackets/Fragment/BracketsFragment.java
package pt.cosmik.boostctrl.external.tournament_brackets.Fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;

import java.util.ArrayList;
import java.util.List;

import pt.cosmik.boostctrl.BoostCtrlApplication;
import pt.cosmik.boostctrl.R;
import pt.cosmik.boostctrl.external.tournament_brackets.adapter.BracketsSectionAdapter;
import pt.cosmik.boostctrl.external.tournament_brackets.model.ColumnData;
import pt.cosmik.boostctrl.external.tournament_brackets.model.CompetitorData;
import pt.cosmik.boostctrl.external.tournament_brackets.model.MatchData;


/**
 * Created by Emil on 21/10/17.
 */

public class BracketsFragment extends Fragment implements ViewPager.OnPageChangeListener {

    private ViewPager viewPager;
    private BracketsSectionAdapter sectionAdapter;
    private ArrayList<ColumnData> sectionList;
    private int mNextSelectedScreen;
    private int mCurrentPagerState;


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_brackets, container, false);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initViews();
        setData();
        initialiseViewPagerAdapter();
    }

    private void setData() {
        sectionList = new ArrayList<>();
        ArrayList<MatchData> Colomn1matchesList = new ArrayList<>();
        ArrayList<MatchData> colomn2MatchesList = new ArrayList<>();
        ArrayList<MatchData> colomn3MatchesList = new ArrayList<>();
        CompetitorData competitorOne = new CompetitorData("Manchester United Fc", "2");
        CompetitorData competitorTwo = new CompetitorData("Arsenal", "1");
        CompetitorData competitorThree = new CompetitorData("Chelsea", "2");
        CompetitorData competitorFour = new CompetitorData("Tottenham", "1");
        CompetitorData competitorFive = new CompetitorData("Manchester FC", "2");
        CompetitorData competitorSix = new CompetitorData("Liverpool", "4");
        CompetitorData competitorSeven = new CompetitorData("West ham ", "2");
        CompetitorData competitorEight = new CompetitorData("Bayern munich", "1");
        MatchData matchData1 = new MatchData(competitorOne,competitorTwo);
        MatchData matchData2 = new MatchData(competitorThree, competitorFour);
        MatchData matchData3 = new MatchData(competitorFive,competitorSix);
        MatchData matchData4 = new MatchData(competitorSeven, competitorEight);
        Colomn1matchesList.add(matchData1);
        Colomn1matchesList.add(matchData2);
        Colomn1matchesList.add(matchData3);
        Colomn1matchesList.add(matchData4);
        ColumnData columnData1 = new ColumnData(Colomn1matchesList);
        sectionList.add(columnData1);
        CompetitorData competitorNine = new CompetitorData("Manchester United Fc", "2");
        CompetitorData competitorTen = new CompetitorData("Chelsea", "4");
        CompetitorData competitorEleven = new CompetitorData("Liverpool", "2");
        CompetitorData competitorTwelve = new CompetitorData("westham", "1");
        MatchData matchData5 = new MatchData(competitorNine,competitorTen);
        MatchData matchData6 = new MatchData(competitorEleven, competitorTwelve);
        colomn2MatchesList.add(matchData5);
        colomn2MatchesList.add(matchData6);
        ColumnData columnData2 = new ColumnData(colomn2MatchesList);
        sectionList.add(columnData2);
        CompetitorData competitorThirteen = new CompetitorData("Chelsea", "2");
        CompetitorData competitorForteen = new CompetitorData("Liverpool", "1");
        MatchData matchData7 = new MatchData(competitorThirteen, competitorForteen);
        colomn3MatchesList.add(matchData7);
        ColumnData columnData3 = new ColumnData(colomn3MatchesList);
        sectionList.add(columnData3);

    }

    private void initialiseViewPagerAdapter() {

        sectionAdapter = new BracketsSectionAdapter(getChildFragmentManager(),this.sectionList);
        viewPager.setOffscreenPageLimit(10);
        viewPager.setAdapter(sectionAdapter);
        viewPager.setCurrentItem(0);
        viewPager.setPageMargin(-200);
        viewPager.setHorizontalFadingEdgeEnabled(true);
        viewPager.setFadingEdgeLength(50);

        viewPager.addOnPageChangeListener(this);
    }

    private void initViews() {
        viewPager = getView().findViewById(R.id.container);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

        if (mCurrentPagerState != ViewPager.SCROLL_STATE_SETTLING) {
            // We are moving to next screen on right side
            if (positionOffset > 0.5) {
                // Closer to next screen than to current
                if (position + 1 != mNextSelectedScreen) {
                    mNextSelectedScreen = position + 1;
                    //update view here
                    if (getBracketsFragment(position).getColumnList().get(0).getHeight()
                            != ((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131))
                        getBracketsFragment(position).shrinkView(((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131));
                    if (getBracketsFragment(position + 1).getColumnList().get(0).getHeight()
                            != ((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131))
                        getBracketsFragment(position + 1).shrinkView(((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131));
                }
            } else {
                // Closer to current screen than to next
                if (position != mNextSelectedScreen) {
                    mNextSelectedScreen = position;
                    //updateViewhere

                    if (getBracketsFragment(position + 1).getCurrentBracketSize() ==
                            getBracketsFragment(position + 1).getPreviousBracketSize()) {
                        getBracketsFragment(position + 1).shrinkView(((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131));
                        getBracketsFragment(position).shrinkView(((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131));
                    } else {
                        int currentFragmentSize = getBracketsFragment(position + 1).getCurrentBracketSize();
                        int previousFragmentSize = getBracketsFragment(position + 1).getPreviousBracketSize();
                        if (currentFragmentSize != previousFragmentSize) {
                            getBracketsFragment(position + 1).expandHeight(((BoostCtrlApplication)getActivity().getApplication()).dpToPx(262));
                            getBracketsFragment(position).shrinkView(((BoostCtrlApplication)getActivity().getApplication()).dpToPx(131));
                        }
                    }
                }
            }
        } else {
            // We are moving to next screen left side
            if (positionOffset > 0.5) {
                // Closer to current screen than to next
                if (position + 1 != mNextSelectedScreen) {
                    mNextSelectedScreen = position + 1;
                    //update view for screen

                }
            } else {
                // Closer to next screen than to current
                if (position != mNextSelectedScreen) {
                    mNextSelectedScreen = position;
                    //updateviewfor screem
                }
            }
        }
    }

    @Override
    public void onPageSelected(int position) {

    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    private BracketsColumnFragment getBracketsFragment(int position) {
        BracketsColumnFragment bracketsFragment = null;
        List<Fragment> fragments = getChildFragmentManager().getFragments();
        for (Fragment fragment : fragments) {
            if (fragment instanceof BracketsColumnFragment) {
                bracketsFragment = (BracketsColumnFragment) fragment;
                if (bracketsFragment.getSectionNumber() == position)
                    break;
            }
        }
        return bracketsFragment;
    }
}

File: app/src/main/java/pt/cosmik/boostctrl/external/fragment/BracketsFragment.kt
package pt.cosmik.boostctrl.external.fragment

import android.app.AlertDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import androidx.viewpager.widget.ViewPager
import com.jakewharton.rxbinding3.view.clicks
import org.koin.android.viewmodel.ext.android.viewModel
import pt.cosmik.boostctrl.BoostCtrlApplication
import pt.cosmik.boostctrl.MainActivity
import pt.cosmik.boostctrl.R
import pt.cosmik.boostctrl.external.adapter.BracketsSectionAdapter
import pt.cosmik.boostctrl.external.model.ColumnData
import pt.cosmik.boostctrl.ui.common.BaseFragment
import pt.cosmik.boostctrl.utils.Constants
import java.util.concurrent.TimeUnit

class BracketsFragment: BaseFragment(), ViewPager.OnPageChangeListener {

    private val vm: BracketsViewModel by viewModel()

    private var viewPager: ViewPager? = null
    private var sectionAdapter: BracketsSectionAdapter? = null
    private var sectionList = arrayListOf<ColumnData>()
    private var mNextSelectedScreen = 0

    private var loadingBar: ProgressBar? = null
    private var containerButton: Button? = null
    private var sectionButton: Button? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_brackets, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        loadingBar = view.findViewById(R.id.loading_bar)
        containerButton = view.findViewById<Button>(R.id.container_button)?.apply {
            disposables.add(clicks().throttleFirst(Constants.THROTTLE_SINGLE_CLICK_MILLISECONDS, TimeUnit.MILLISECONDS).subscribe {
                vm.processEvent(BracketsViewModel.BracketsFragmentEvent.SelectedChangeContainerButton)
            })
        }
        sectionButton = view.findViewById<Button>(R.id.section_button)?.apply {
            disposables.add(clicks().throttleFirst(Constants.THROTTLE_SINGLE_CLICK_MILLISECONDS, TimeUnit.MILLISECONDS).subscribe {
                vm.processEvent(BracketsViewModel.BracketsFragmentEvent.SelectedChangeSectionButton)
            })
        }

        vm.viewState.observe(this, Observer {
            if (it.isLoading) {
                loadingBar?.visibility = View.VISIBLE
                containerButton?.isEnabled = false
                sectionButton?.isEnabled = false
            }
            else {
                loadingBar?.visibility = View.GONE
                containerButton?.isEnabled = true
                sectionButton?.isEnabled = true
            }
        })

        vm.viewEffect.observe(this, Observer {
            when (it) {
                is BracketsViewModel.BracketsFragmentViewEffect.ShowError -> showErrorMessage(it.error)
                is BracketsViewModel.BracketsFragmentViewEffect.ShowBrackets -> {
                    sectionList = it.brackets
                    initialiseViewPagerAdapter()
                }
                BracketsViewModel.BracketsFragmentViewEffect.ShowNoBracketsAvailable -> {
                    showErrorMessage("There are no brackets available for this competition")
                    findNavController().popBackStack()
                }
                is BracketsViewModel.BracketsFragmentViewEffect.ShowDialogContainerPicker -> {
                    val builder = AlertDialog.Builder(context)
                    builder.setTitle(context?.getString(R.string.select))
                        .setItems(it.containers.toTypedArray()) { _, which ->
                            containerButton?.text = it.containers[which]
                            sectionButton?.text = context?.getString(R.string.chose_section)
                            vm.processEvent(BracketsViewModel.BracketsFragmentEvent.SelectedContainer(it.containers[which]))
                        }
                    builder.create().show()

                }
                is BracketsViewModel.BracketsFragmentViewEffect.ShowDialogSectionPicker -> {
                    val builder = AlertDialog.Builder(context)
                    builder.setTitle(context?.getString(R.string.select))
                        .setItems(it.sections.toTypedArray()) { _, which ->
                            sectionButton?.text = it.sections[which]
                            vm.processEvent(BracketsViewModel.BracketsFragmentEvent.SelectedSection(it.sections[which]))
                        }
                    builder.create().show()
                }
            }
        })

        vm.processEvent(BracketsViewModel.BracketsFragmentEvent.ViewCreated(arguments?.get("competitionId") as? String))
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initViews()
    }

    private fun initialiseViewPagerAdapter() {
        if (sectionAdapter != null) {
            viewPager!!.adapter = BracketsSectionAdapter(childFragmentManager, sectionList)
        }
        else {
            sectionAdapter = BracketsSectionAdapter(childFragmentManager, sectionList)
            viewPager!!.offscreenPageLimit = 10
            viewPager!!.adapter = sectionAdapter
            viewPager!!.currentItem = 0
            viewPager!!.pageMargin = -200
            viewPager!!.isHorizontalFadingEdgeEnabled = true
            viewPager!!.setFadingEdgeLength(50)
            viewPager!!.addOnPageChangeListener(this)
        }
    }

    private fun initViews() {
        viewPager = view!!.findViewById(R.id.container)
    }

    override fun showErrorMessage(message: String) {
        (activity as MainActivity).showMessageInSnackBar(message)
    }

    override fun getActionBarTitle(): String = "title"

    override fun onDestroyView() {
        super.onDestroyView()
        removeObservers()
    }

    override fun removeObservers() {
        vm.viewState.removeObservers(this)
        vm.viewEffect.removeObservers(this)
    }

    override fun onPageScrolled(
        position: Int,
        positionOffset: Float,
        positionOffsetPixels: Int
    ) {
        if (positionOffset > 0.5) { // Closer to next screen than to current
            if (position + 1 != mNextSelectedScreen) {
                mNextSelectedScreen = position + 1
                //update view here
                if (getBracketsFragment(position)!!.columnList[0].height
                    != (activity!!.application as BoostCtrlApplication).dpToPx(
                        131
                    )
                ) getBracketsFragment(position)!!.shrinkView(
                    (activity!!.application as BoostCtrlApplication).dpToPx(
                        131
                    )
                )
                if (getBracketsFragment(position + 1)!!.columnList[0].height
                    != (activity!!.application as BoostCtrlApplication).dpToPx(
                        131
                    )
                ) getBracketsFragment(position + 1)!!.shrinkView(
                    (activity!!.application as BoostCtrlApplication).dpToPx(
                        131
                    )
                )
            }
        } else { // Closer to current screen than to next
            if (position != mNextSelectedScreen) {
                mNextSelectedScreen = position
                //update View here
                if (getBracketsFragment(position + 1)!!.currentBracketSize ==
                    getBracketsFragment(position + 1)!!.previousBracketSize
                ) {
                    getBracketsFragment(position + 1)!!.shrinkView(
                        (activity!!.application as BoostCtrlApplication).dpToPx(
                            131
                        )
                    )
                    getBracketsFragment(position)!!.shrinkView(
                        (activity!!.application as BoostCtrlApplication).dpToPx(
                            131
                        )
                    )
                } else {
                    val currentFragmentSize: Int =
                        getBracketsFragment(position + 1)!!.currentBracketSize
                    val previousFragmentSize: Int =
                        getBracketsFragment(position + 1)!!.previousBracketSize
                    if (currentFragmentSize != previousFragmentSize) {
                        getBracketsFragment(position + 1)!!.expandHeight(
                            (activity!!.application as BoostCtrlApplication).dpToPx(
                                262
                            )
                        )
                        getBracketsFragment(position)!!.shrinkView(
                            (activity!!.application as BoostCtrlApplication).dpToPx(
                                131
                            )
                        )
                    }
                }
            }
        }
    }

    override fun onPageSelected(position: Int) {}

    override fun onPageScrollStateChanged(state: Int) {}

    private fun getBracketsFragment(position: Int): BracketsColumnFragment? {
        var bracketsFragment: BracketsColumnFragment? = null
        val fragments = childFragmentManager.fragments
        for (fragment in fragments) {
            if (fragment is BracketsColumnFragment) {
                bracketsFragment = fragment
                if (bracketsFragment.sectionNumber == position) break
            }
        }
        return bracketsFragment
    }
}